internal/fwserver/server_readdatasource.go

```diff
@@ -38,6 +38,13 @@ func (s *Server) ReadDataSource(ctx context.Context, req *ReadDataSourceRequest,
 		return
 	}
 
+	//TODO: add logging for replacing deferred reason
+	if s.deferred != nil {
+		resp.Deferred = &datasource.Deferred{
+			Reason: resp.Deferred.Reason,
+		}
+	}
```

Comment:
See comment in `ReadResource`, same idea applies.

SDKv2 logic: https://github.com/hashicorp/terraform-plugin-sdk/blob/e90e8a2497cb1bced1d5614e65c904453244d18c/helper/schema/grpc_provider.go#L1374-L1398

Your logic for constructing an unknown value will be a little different since it's not using `cty`, but the idea is the same, we'll set `State` to an unknown value for the entire data source object.